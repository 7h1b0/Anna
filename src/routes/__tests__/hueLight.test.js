const request = require('supertest');
const knex = require('../../knexClient');
const User = require('../../modules/models/user');
const hueLight = require('../../modules/models/hueLight');
const app = require('../../index.js');

jest.mock('../../services/requestService');
const requestService = require('../../services/requestService');

const user = {
  user_id: 1,
  username: 'test',
  password: '$2a$10$4ftuQxquI/5NR3POJy.2O.DmscxoSdCBzUvlnX2iXGMxtpqhd3w6O', // anna
  token: '8e6a76928f76d23665f78ff3688ca86422d5',
};

const hueLightRooms = [
  {
    light_id: 1,
    room_id: 1,
  },
  {
    light_id: 2,
    room_id: 1,
  },
];

describe('Hue Light API', () => {
  beforeAll(async () => {
    await knex(User.TABLE).insert(user);
    await knex(hueLight.TABLE).insert(hueLightRooms);
  });

  afterEach(() => {
    requestService.put.mockClear();
    requestService.get.mockClear();
  });

  afterAll(async () => {
    await knex(User.TABLE).truncate();
    await knex(hueLight.TABLE).truncate();
  });

  describe('/api/hue/lights', () => {
    describe('GET', () => {
      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get('/api/hue/lights')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });

      it('should return all lights', async () => {
        const response = await request(app)
          .get('/api/hue/lights')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(200);
        expect(response.body).toMatchSnapshot();
      });
    });
  });

  describe('api/hue/lights/:id_light', () => {
    describe('GET', () => {
      it('should return a light', async () => {
        const response = await request(app)
          .get('/api/hue/lights/2')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(200);
        expect(response.body).toMatchSnapshot();
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get('/api/hue/lights/2')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });
    });

    describe('PATCH', () => {
      it('should update the name of the light', async () => {
        const updatedName = { name: 'test' };

        const response = await request(app)
          .patch('/api/hue/lights/2')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send(updatedName);

        expect(response.status).toBe(200);
        expect(requestService.put).toHaveBeenCalledTimes(1);
        expect(requestService.put.mock.calls[0][1]).toEqual(updatedName);
      });

      it('should update the roomId and the name', async () => {
        const updatedName = { roomId: 2, name: 'test_room' };

        const response = await request(app)
          .patch('/api/hue/lights/2')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send(updatedName);

        expect(response.status).toBe(200);
        expect(requestService.put).toHaveBeenCalledTimes(1);
        expect(requestService.put.mock.calls[0][1]).toEqual({
          name: 'test_room',
        });

        const res = await knex(hueLight.TABLE)
          .select('room_id')
          .where('light_id', 2);

        expect(res).toEqual([{ room_id: 2 }]);
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .patch('/api/hue/lights/2')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });

      it('should retun 400 if name is missing', async () => {
        const updatedName = { test: 'test' };

        const response = await request(app)
          .patch('/api/hue/lights/2')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send(updatedName);

        expect(response.status).toBe(400);
      });
    });
  });

  describe('api/hue/lights/:id_light/:status', () => {
    it('should retun 401 when user is not authenticated', async () => {
      const response = await request(app)
        .get('/api/hue/lights/2/on')
        .set('Accept', 'application/json')
        .set('x-access-token', 'fake');

      expect(response.status).toBe(401);
    });

    it('should call requestService with on to true', async () => {
      const response = await request(app)
        .get('/api/hue/lights/2/on')
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toBe(200);
      expect(requestService.put).toHaveBeenCalledTimes(1);
      expect(requestService.put.mock.calls[0][1]).toEqual({ on: true });
      expect(requestService.put.mock.calls[0][0]).toMatch(/lights\/2\/state$/);
    });

    it('should call requestService with on to false', async () => {
      const response = await request(app)
        .get('/api/hue/lights/2/off')
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toBe(200);
      expect(requestService.put).toHaveBeenCalledTimes(1);
      expect(requestService.put.mock.calls[0][1]).toEqual({ on: false });
      expect(requestService.put.mock.calls[0][0]).toMatch(/lights\/2\/state$/);
    });
  });

  describe('api/hue/lights/:id_light/state', () => {
    it('should retun 401 when user is not authenticated', async () => {
      const response = await request(app)
        .patch('/api/hue/lights/2/state')
        .set('Accept', 'application/json')
        .set('x-access-token', 'fake');

      expect(response.status).toBe(401);
    });

    it('should return 400 when body is invalid', async () => {
      const response = await request(app)
        .patch('/api/hue/lights/2/state')
        .set('Accept', 'application/json')
        .set('x-access-token', user.token)
        .send({});

      expect(response.status).toBe(400);
    });

    it('should return 400 when body is invalid', async () => {
      const response = await request(app)
        .patch('/api/hue/lights/2/state')
        .set('Accept', 'application/json')
        .set('x-access-token', user.token)
        .send({ test: true });

      expect(response.status).toBe(400);
    });

    it('should call requestService with body passed', async () => {
      const body = { sat: 200, bri: 25, on: true, xy: [23, 45] };
      const response = await request(app)
        .patch('/api/hue/lights/2/state')
        .set('Accept', 'application/json')
        .set('x-access-token', user.token)
        .send(body);

      expect(response.status).toBe(200);
      expect(requestService.put).toHaveBeenCalledTimes(1);
      expect(requestService.put.mock.calls[0][1]).toEqual(body);
      expect(requestService.put.mock.calls[0][0]).toMatch(/lights\/2\/state$/);
    });
  });
});
