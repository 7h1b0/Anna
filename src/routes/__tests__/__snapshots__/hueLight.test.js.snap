// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hue Light API /api/hue/lights GET should return all lights 1`] = `
Array [
  Object {
    "id": "1",
    "manufacturername": "Philips",
    "modelid": "LCT001",
    "name": "MOCK",
    "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "state": Object {
      "alert": "none",
      "bri": 163,
      "colormode": "xy",
      "ct": 272,
      "effect": "none",
      "hue": 15736,
      "on": true,
      "reachable": true,
      "sat": 32,
      "xy": Array [
        0.3969,
        0.3851,
      ],
    },
    "type": "Extended color light",
  },
  Object {
    "id": "2",
    "manufacturername": "Philips",
    "modelid": "LCT001",
    "name": "JEST",
    "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b3",
    "state": Object {
      "alert": "none",
      "bri": 97,
      "colormode": "xy",
      "ct": 272,
      "effect": "none",
      "hue": 17161,
      "on": false,
      "reachable": true,
      "sat": 38,
      "xy": Array [
        0.3978,
        0.3883,
      ],
    },
    "type": "Extended color light",
  },
]
`;

exports[`Hue Light API api/hue/lights/:id_light GET should return a light 1`] = `
Object {
  "manufacturername": "Philips",
  "modelid": "LCT001",
  "name": "MOCK",
  "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b3",
  "state": Object {
    "alert": "none",
    "bri": 163,
    "colormode": "xy",
    "ct": 272,
    "effect": "none",
    "hue": 15736,
    "on": false,
    "reachable": true,
    "sat": 32,
    "xy": Array [
      0.3969,
      0.3851,
    ],
  },
  "type": "Extended color light",
}
`;

exports[`Hue Light API api/hue/lights/:id_light PATCH should update the name of the light 1`] = `
Object {
  "body": Object {
    "name": "test",
  },
}
`;

exports[`Hue Light API api/hue/lights/:id_light PATCH should update the roomId and the name 1`] = `
Object {
  "body": Object {
    "name": "test_room",
  },
}
`;

exports[`Hue Light API api/hue/lights/:id_light/:status should call requestService with on to false 1`] = `
Object {
  "on": false,
}
`;

exports[`Hue Light API api/hue/lights/:id_light/:status should call requestService with on to true 1`] = `
Object {
  "on": true,
}
`;

exports[`Hue Light API api/hue/lights/:id_light/state should call requestService with body passed 1`] = `
Object {
  "bri": 25,
  "on": true,
  "sat": 200,
  "xy": Array [
    23,
    45,
  ],
}
`;
